### ISSUE TITLE
add `which` to the docker image

### ISSUE THREAD

**Issue by cmatKhan** [nextflow-io/nextflow#4219]
### Expected behavior and actual behavior

I am trying to use the nextflow/nextflow docker image on a cluster that uses the LSF scheduler. This particular cluster (not sure if this is a LSF feature) requires that all computation is performed in a docker container. I cannot install nextflow directly -- i have to run it in a container.  

I'd like to use the nextflow/nextflow dockerhub image. But, `which` is not present in the container.

### Steps to reproduce the problem

```bash
docker run -it --rm nextflow/nextflow /bin/bash
bash-4.2# which
bash: which: command not found
```

### Environment 

Current dockerhub nextflow container

### Request

Add `which` to the container

**Comment by bentsherman** [comment]
Sorry, but we try not to add any unnecessary dependencies to the docker image in order to keep it as small as possible. Feel free to make your own image with whatever extra tools you need.

**Comment by cmatKhan** [comment]
@bentsherman 

`which` is used in the nextflow script. How is this an unnecessary dependency?

https://github.com/nextflow-io/nextflow/blob/80c5cb276c6d1857897b7883ef7f01deea8f635c/nextflow#L256

**Comment by bentsherman** [comment]
That code path isn't typically used in the docker image. Are you trying to run it with `NXF_PACK=all` ?

**Comment by cmatKhan** [comment]
I do see what you mean re: `NXF_PACK=all`.

so maybe it is this line that I am encountering on the cluster:

https://github.com/nextflow-io/nextflow/blob/80c5cb276c6d1857897b7883ef7f01deea8f635c/nextflow#L277

**Comment by bentsherman** [comment]
Try the following command: `NXF_DEBUG='x' nextflow run hello`

That will echo every command that is executed so that we can see exactly where it fails

**Comment by cmatKhan** [comment]
hm, right -- i should have turned the debug log on to begin with. thank you.

So this is what happens without turning the log on:

```
...
/usr/local/bin/nextflow: line 277: which: command not found
/usr/local/bin/nextflow: line 154: /scratch1/fs1/doering/chasem/nextflow/capsule/apps/nextflow_23.08.1-edge/java: No such file or directory
```

But, with the `NXF_DEBUG=x` log turned on, I see that at one point this is set:

```
...
+ for x in '"${launcher[@]}"'
+ x='"-Dcapsule.jar=/scratch1/fs1/doering/chasem/nextflow/framework/23.08.1-edge/nextflow-23.08.1-edge-one.jar'
+ x=-Dcapsule.jar=/scratch1/fs1/doering/chasem/nextflow/framework/23.08.1-edge/nextflow-23.08.1-edge-one.jar
...
```

But then by the last lines of that file, I see:

```
...
 exec /scratch1/fs1/doering/chasem/nextflow/capsule/apps/nextflow_23.08.1-edge/java --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent.locks=ALL-UNNAMED --add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED
...
/usr/local/bin/nextflow: line 154: /scratch1/fs1/doering/chasem/nextflow/capsule/apps/nextflow_23.08.1-edge/java: No such file or directory
```

In the `NXF_HOME` dir, this does exist:

```
nextflow/framework/23.08.1-edge/nextflow-23.08.1-edge-one.jar
```
but this is empty:
```
nextflow/capsule/apps/nextflow_23.08.1-edge/
```
This is beyond my knowledge of what NF is doing at this level, but it looks like there is some sort of PATH redirection happening. I am also sure that there is something idiosyncratic about the cluster I am trying to use right now -- these issues abound. It is not just NF.

I'm not sure that having `which` would solve the problem, but I can say that on this system, I am hitting *a line* that is trying to use `which`, and it does not exist

**Comment by bentsherman** [comment]
I'm confused by what's happening here. If `which java` fails then it falls back to `JAVA_CMD=java`, but somehow it ends up using `/scratch1/fs1/doering/chasem/nextflow/capsule/apps/nextflow_23.08.1-edge/java` which does not exist. But I don't see how JAVA_CMD is changed to that...

Well, you can at least test it with which pretty easily. Just run `yum install which` in the container, then run Nextflow.

**Comment by pditommaso** [comment]
Solved by 817aa05bb
